### PR TITLE
Remove unused move to allow RVO works

### PR DESCRIPTION
**Pull request by mapleFU**
We don't need to `return move(unique_ptr)`, if the pointer is created inside the function. It will disallow the compiler using copy elision to optimize functions.

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR!

Did you check if this passed the CI on your fork? The reason those moves were there IIRC is that older GCC versions (e.g. GCC-4.8) do require this move to be there if the type is inherited (e.g. if we return a `PhysicalChunkScan` instead of a `PhysicalOperator`).

**Comment by mapleFU** [comment]
@Mytherin I'm using clang11 and gcc8.3, and it's no problem.

The gcc bug seems like: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70972. It violates the C++ standard. Maybe I can using a macro like:

```
#if it's gcc 4.8 or other
  using std::move
#else
  define move as nop
#endif
```

Would it be ok if my current change cannot compile under gcc4.8? @Mytherin

**Comment by Mytherin** [comment]
> The gcc bug seems like: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70972. It violates the C++ standard. Maybe I can using a macro like:

I don't think it is that bug. 

> Would it be ok if my current change cannot compile under gcc4.8?

Compiling under GCC4.8 is unfortunately required since we compile with old versions of GCC to generate binaries that depend on old versions of GLIBC, which is necessary to generate portable binaries. Perhaps there is a better way of generating those binaries, but the CI jobs would have to be rewritten/replaced for that to happen. Until then we can't remove the CI jobs that run with GCC4.8.

**Comment by mapleFU** [comment]
> > Would it be ok if my current change cannot compile under gcc4.8?

Got it, I would like to write a macro like: 

```
#if !defined(__clang__) && defined(__GNUC__) && (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 < 40900)
 using move_return = move;
#else
 #define move_return(v) v
#endif
```

Here I define `return move()` as `move_return`, when compiler need to move, I just using `std::move`, otherwise, it's just a nop. @Mytherin

**Comment by Mytherin** [comment]
> Here I define return move() as move_return, when compiler need to move, I just using std::move, otherwise, it's just a nop. @Mytherin

That's fine to me, to points:

1. Please figure out more precisely which version of GCC requires this (e.g. does this work on V5, V6, etc)
2. Please prefix the macro with `D_` to avoid potential conflicts with other libraries, e.g. name it `D_MOVE_RETURN`.

**Comment by mapleFU** [comment]
@Mytherin OK. I'm busy these two days. I will finish this in this weekend.

**Comment by hannesmuehleisen** [comment]
@mapleFU could you perhaps merge this?

**Comment by mapleFU** [comment]
> @mapleFU could you perhaps merge this?

Sorry, I forgot it because Chinese Spring Festival, so sorry for that. After talk with my friend @frederick-vs-ja, I found that it's   
https://wg21.cmeerw.net/cwg/issue1579 . And it's need some time to check it for compiler like MSVC.

**Comment by frederick-vs-ja** [comment]
>```C++
> #if !defined(__clang__) && defined(__GNUC__) && (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 < 40900)
>  using move_return = move;
> #else
>  #define move_return(v) v
> #endif
>```

We should use `50000` in the condition.

All of the problematic cases are returning a `std::unique_ptr<Derived>` variable when the return type is `std::unique_ptr<Base>`.
Such usage is allowed by post-C++11 defect report [WG21-CWG1579](https://wg21.link/cwg1579), and available since gcc 5/clang 3.9/icc 18/msvc cl 19.0 (VS 2015).

File: src/optimizer/statistics/expression/propagate_cast.cpp [hunk 1]
```diff
@@ -17,7 +17,7 @@ static unique_ptr<BaseStatistics> StatisticsOperationsNumericNumericCast(const B
 	if (input.validity_stats) {
 		stats->validity_stats = input.validity_stats->Copy();
 	}
-	return move(stats);
+	return stats;
```
Comment (frederick-vs-ja) [inline src/optimizer/statistics/expression/propagate_cast.cpp:20]: Should be reverted.

File: src/optimizer/statistics/expression/propagate_constant.cpp [hunk 2]
```diff
@@ -23,7 +23,7 @@ unique_ptr<BaseStatistics> StatisticsPropagator::StatisticsFromValue(const Value
 	case PhysicalType::DOUBLE: {
 		auto result = make_unique<NumericStatistics>(input.type(), input, input);
 		result->validity_stats = make_unique<ValidityStatistics>(input.is_null, !input.is_null);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/optimizer/statistics/expression/propagate_constant.cpp:26]: Should be reverted.

File: src/optimizer/statistics/expression/propagate_constant.cpp [hunk 3]
```diff
@@ -47,7 +47,7 @@ unique_ptr<BaseStatistics> StatisticsPropagator::StatisticsFromValue(const Value
 				result->child_stats[i] = StatisticsFromValue(input.struct_value[i]);
 			}
 		}
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/optimizer/statistics/expression/propagate_constant.cpp:50]: Should be reverted.

File: src/optimizer/statistics/expression/propagate_constant.cpp [hunk 4]
```diff
@@ -64,7 +64,7 @@ unique_ptr<BaseStatistics> StatisticsPropagator::StatisticsFromValue(const Value
 				}
 			}
 		}
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/optimizer/statistics/expression/propagate_constant.cpp:67]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_view_info.hpp [hunk 5]
```diff
@@ -36,7 +36,7 @@ struct CreateViewInfo : public CreateInfo {
 		result->aliases = aliases;
 		result->types = types;
 		result->query = unique_ptr_cast<SQLStatement, SelectStatement>(query->Copy());
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_view_info.hpp:39]: Should be reverted.

File: src/include/duckdb/optimizer/filter_combiner.hpp [hunk 6]
```diff
@@ -83,7 +83,7 @@ class FilterCombiner {
 			auto const_filter = make_unique<ConstantFilter>(comp_expr->type, const_value);
 			conj_filter->child_filters.push_back(move(const_filter));
 		}
-		return move(conj_filter);
+		return conj_filter;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/optimizer/filter_combiner.hpp:86]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_sequence_info.hpp [hunk 7]
```diff
@@ -46,7 +46,7 @@ struct CreateSequenceInfo : public CreateInfo {
 		result->max_value = max_value;
 		result->start_value = start_value;
 		result->cycle = cycle;
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_sequence_info.hpp:49]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_scalar_function_info.hpp [hunk 8]
```diff
@@ -36,7 +36,7 @@ struct CreateScalarFunctionInfo : public CreateFunctionInfo {
 		set.functions = functions;
 		auto result = make_unique<CreateScalarFunctionInfo>(move(set));
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_scalar_function_info.hpp:39]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_table_info.hpp [hunk 9]
```diff
@@ -44,7 +44,7 @@ struct CreateTableInfo : public CreateInfo {
 		if (query) {
 			result->query = unique_ptr_cast<SQLStatement, SelectStatement>(query->Copy());
 		}
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_table_info.hpp:47]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_aggregate_function_info.hpp [hunk 10]
```diff
@@ -34,7 +34,7 @@ struct CreateAggregateFunctionInfo : public CreateFunctionInfo {
 	unique_ptr<CreateInfo> Copy() const override {
 		auto result = make_unique<CreateAggregateFunctionInfo>(functions);
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_aggregate_function_info.hpp:37]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_collation_info.hpp [hunk 11]
```diff
@@ -35,7 +35,7 @@ struct CreateCollationInfo : public CreateInfo {
 	unique_ptr<CreateInfo> Copy() const override {
 		auto result = make_unique<CreateCollationInfo>(name, function, combinable, not_required_for_equality);
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_collation_info.hpp:38]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_copy_function_info.hpp [hunk 12]
```diff
@@ -28,7 +28,7 @@ struct CreateCopyFunctionInfo : public CreateInfo {
 	unique_ptr<CreateInfo> Copy() const override {
 		auto result = make_unique<CreateCopyFunctionInfo>(function);
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_copy_function_info.hpp:31]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_index_info.hpp [hunk 13]
```diff
@@ -42,7 +42,7 @@ struct CreateIndexInfo : public CreateInfo {
 		for (auto &expr : expressions) {
 			result->expressions.push_back(expr->Copy());
 		}
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_index_info.hpp:45]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_macro_info.hpp [hunk 14]
```diff
@@ -25,7 +25,7 @@ struct CreateMacroInfo : public CreateFunctionInfo {
 		result->function = function->Copy();
 		result->name = name;
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_macro_info.hpp:28]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_pragma_function_info.hpp [hunk 15]
```diff
@@ -33,7 +33,7 @@ struct CreatePragmaFunctionInfo : public CreateFunctionInfo {
 	unique_ptr<CreateInfo> Copy() const override {
 		auto result = make_unique<CreatePragmaFunctionInfo>(functions[0].name, functions);
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_pragma_function_info.hpp:36]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_schema_info.hpp [hunk 16]
```diff
@@ -20,7 +20,7 @@ struct CreateSchemaInfo : public CreateInfo {
 	unique_ptr<CreateInfo> Copy() const override {
 		auto result = make_unique<CreateSchemaInfo>();
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_schema_info.hpp:23]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_table_function_info.hpp [hunk 17]
```diff
@@ -32,7 +32,7 @@ struct CreateTableFunctionInfo : public CreateFunctionInfo {
 		set.functions = functions;
 		auto result = make_unique<CreateTableFunctionInfo>(move(set));
 		CopyProperties(*result);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_table_function_info.hpp:35]: Should be reverted.

File: src/include/duckdb/parser/parsed_data/create_type_info.hpp [hunk 18]
```diff
@@ -31,7 +31,7 @@ struct CreateTypeInfo : public CreateInfo {
 		CopyProperties(*result);
 		result->name = name;
 		result->type = make_unique<LogicalType>(*type);
-		return move(result);
+		return result;
```
Comment (frederick-vs-ja) [inline src/include/duckdb/parser/parsed_data/create_type_info.hpp:34]: Should be reverted.

**Review by frederick-vs-ja — COMMENTED** [review]
There are still some implicit moves from `std::unique_ptr<Derived>` to `std::unique_ptr<Base>`. I don't know if they are OK.

**Comment by mapleFU** [comment]
I'll create another pr for this task.

**Comment by frederick-vs-ja** [comment]
I've roughly looked into this. But it seems that on 2024-05-20 almost all instances of the `return std::move(LocalVariable);` pattern are returning a `unique_ptr<DerivedClass>` with return type being `unique_ptr<BaseClass>`.

548bab5ac355a981f562f79ae34e78f6d0d18044 #8050 removed one unnecessary occurrence of `std::move`.

I don't know whether there's now any occurrence of `std::move` where NRVO can work with `std::move` removed. If there's no, this task can be considered completed.

**Pull request closed without merging**